pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+		#[pallet::weight(T::Weights::fill_order_full())]
+		pub fn fill_order_full(origin: OriginFor<T>, order_id: T::OrderIdNonce) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			let order = <Orders<T>>::get(order_id)?;
+			// maybe move to ensure if we don't need these later
+			// might need decimals from currency, but should hopefully be able to use FP
+			// price/amounts from FP balance
+
+			let sell_amount = order.buy_amount.ensure_mul(order.price)?;
+
+			ensure!(
+				T::TradeableAsset::can_reserve(order.asset_in_id, &account_id, order.buy_amount),
+				Error::<T>::InsufficientAssetFunds,
+			);
+
+			Self::unreserve_order(&order)?;
+			T::TradeableAsset::transfer(
+				order.asset_in_id,
+				&account_id,
+				&order.placing_account,
+				order.buy_amount,
+			)?;
+			T::TradeableAsset::transfer(
+				order.asset_out_id,
+				&order.placing_account,
+				&account_id,
+				sell_amount,
+			)?;
+			Self::remove_order(order.order_id)?;
+			Self::deposit_event(Event::OrderFulfillment {
+				order_id,
+				placing_account: order.placing_account,
+				fulfilling_account: account_id,
+				partial_fulfillment: true,
+				currency_in: order.asset_in_id,
+				currency_out: order.asset_out_id,
+				fulfillment_amount: order.buy_amount,
+				sell_price_limit: order.price,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Remove an order from storage
+		pub fn remove_order(order_id: T::OrderIdNonce) -> DispatchResult {
+			let order = <Orders<T>>::get(order_id)?;
+			<UserOrders<T>>::remove(&order.placing_account, order.order_id);
+			<Orders<T>>::remove(order.order_id);
+			let mut orders = <AssetPairOrders<T>>::get(order.asset_in_id, order.asset_out_id);
+			orders.retain(|o| *o != order.order_id);
+			<AssetPairOrders<T>>::insert(order.asset_in_id, order.asset_out_id, orders);
+			Ok(())
+		}
+
+		/// Get reserve amount when fee and out currency are the same
+		pub fn get_combined_reserve(
+			sell_amount: T::ForeignCurrencyBalance,
+		) -> Result<FeeBalance<T>, DispatchError> {
+			let fee_amount = T::Fees::fee_value(T::OrderFeeKey::get());
+
+			let sell_reserve_balance: FeeBalance<T> = sell_amount
+				.try_into()
+				.map_err(|_| Error::<T>::BalanceConversionErr)?;
+			Ok(sell_reserve_balance.ensure_add(fee_amount)?)
+		}
+
+		/// Unreserve funds for an order that is finished either
+		/// through fulfillment or cancellation.
+		pub fn unreserve_order(order: &OrderOf<T>) -> Result<(), DispatchError> {
+			if T::FeeCurrencyId::get() == order.asset_out_id {
+				let total_reserve_amount = Self::get_combined_reserve(order.max_sell_amount)?;
+				T::ReserveCurrency::unreserve(&order.placing_account, total_reserve_amount);
+			} else {
+				T::TradeableAsset::unreserve(
+					order.asset_out_id,
+					&order.placing_account,
+					order.max_sell_amount,
+				);
+				T::ReserveCurrency::unreserve(
+					&order.placing_account,
+					T::Fees::fee_value(T::OrderFeeKey::get()),
+				);
+			};
+			Ok(())
+		}
```

Comment:
>I was wondering how much work it would be to make fees also usable for handling deposits at a centralised location in our system

Could you expand this? Not sure what you mean by a centralized location

Reply:
One pallet doing reserves. Either variable or with key values as for fees. For each account. Making it easy to follow for which reasons the reserve happened if we add an identifier to the different amounts or so.

Reply:
So just what `pallet_fees` does, but for reserves, right? 
`pallet_fees` has methods as `fee_to_author/fee_to_treasury`/... You may say something like: `fee_to_reserve(account, fee_key)`?